### PR TITLE
added Id class function for EoL

### PR DESCRIPTION
**Pull request by Daniel-Davies**
Added the Encylopedia of Life functions to obtain IDs of given species. Also refactored tests.

## Related Issue
#62 

## Example
c = Ids('panthera tigris')
c.eol()
c.extract_ids()

### CODE REVIEW

**Comment by sckott** [comment]
So some code needs to be re-worked. The EOL API url you used results in "pages" which do have IDs, but are page ids, not taxon ids. To get taxon ids you need to call e.g., https://eol.org/api/pages/1.0/328674.json 

In the R get_eolid fxn, I get pages here https://github.com/ropensci/taxize/blob/master/R/get_eolid.R#L143 then get taxon ids for all those pages here https://github.com/ropensci/taxize/blob/master/R/get_eolid.R#L161 - - where e.g., the first page id in the result for your example above is https://eol.org/api/pages/1.0/328674.json - which has taxon name, rank name, etc. - which also sorts out the other issue I commented on that the `_make_id` call was missing taxon rank (and name too)

**Comment by Daniel-Davies** [comment]
I see what you mean. Do you think it's appropriate to limit the queries in some way here? It seems that page IDs for many names generate hundreds of IDs. Do you think we should take e.g. first 10 results, or it is better to pass everything back to the user?

**Comment by sckott** [comment]
In R taxize we [limit result of the pages search to names that match the users query (via regex)](https://github.com/ropensci/taxize/blob/master/R/get_eolid.R#L152) - and then fetch the taxon data for each taxon at the `/pages/1.0/xxxxx.json` routes - does that make sense to do here?

**Comment by Daniel-Davies** [comment]
I hope this most recent commit fixes it. It filters the page ids in a similar way to your use of grep in R taxize, and then translates the page ids to the taxa ids.

File: pytaxize/ids/eol_helpers.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        .json()
+        .get("results", {})
+    )
+    return response
+
+
+def process_eol_response(list_of_response_dicts):
+    extracted_ids = list(
+        map(
+            lambda x: _make_id(x.get("id", None), None, None, "eol"),
```
[the remark this answers is not in the document]
Comment (sckott) [reply pytaxize/ids/eol_helpers.py]: Looks like you have a uri now, but you're using the template `https://eol.org/api/search/1.0.json?q=%s&exact=true`. Let's use `https://eol.org/pages/<pageid>/`

File: pytaxize/ids/format_helpers.py [hunk 2]
```diff
@@ -23,6 +23,10 @@ def _make_id(id_, name, rank, type_):
         "species": "https://api.gbif.org/v1/species/suggest?q=%s&rank=%s",
         "other": "https://api.gbif.org/v1/species/suggest?q=%s&rank=%s",
     },
+    "eol": {
+        "species": "https://eol.org/api/search/1.0.json?q=%s&exact=true",
```
[the remark this answers is not in the document]
Comment (sckott) [reply pytaxize/ids/format_helpers.py:27]: Also, in R `taxize::get_eolid` we give back the page id and the taxon id, e.g., 

```r
get_eolid(sci_com='Pinus contorta')
#> [1] "6946713"
#> attr(,"class")
#> [1] "eolid"
#> attr(,"pageid")
#> [1] "1061758"
#> attr(,"provider")
#> [1] "EOL Dynamic Hierarchy 0.9"
#> attr(,"match")
#> [1] "found"
#> attr(,"multiple_matches")
#> [1] TRUE
#> attr(,"pattern_match")
#> [1] FALSE
#> attr(,"uri")
#> [1] "https://eol.org/pages/1061758/"
```

[the remark this answers is not in the document]
Comment (Daniel-Davies) [reply pytaxize/ids/format_helpers.py:27]: Sure I can do that- and I'll change the URL to give back the page URL also

File: pytaxize/ids/eol_helpers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Daniel-Davies) [reply pytaxize/ids/eol_helpers.py]: The most recent commit should have this changed- do let me know if I've missed something however

File: pytaxize/ids/eol_helpers.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
+from ..refactor import Refactor
+from .format_helpers import _make_id
+
+
+def eol_query_for_single_name(name):
+    # will replace later with either a pre-made package
+    # or with an implementation of EoL in taxize
+    response = (
+        Refactor(
+            url="https://eol.org/api/search/1.0.json?q=%s&exact=true" % name,
```
[the remark this answers is not in the document]
Comment (sckott) [reply pytaxize/ids/eol_helpers.py:10]: @Daniel-Davies did you see this comment?

[the remark this answers is not in the document]
Comment (Daniel-Davies) [reply pytaxize/ids/eol_helpers.py]: My bad! Should be fixed now

**Pull request merged**